### PR TITLE
Add IR transform support

### PR DESCRIPTION
**Pull request by MikeStall**
Support for IR rewriting.  Since IR is internal, these changes are internal too and not directly exposed. 
This enables things like Delegation and constant folding. 

IR Rewriters can be registered with the engine and then run on ApplyErrors and ApplyIR.

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/IR/Visitors/RewritingIRVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    // Common helper for IR Rewriting. 
+    // IR trees are immutable, so we need to walk and rewrite the tree to include any new nodes.
+    // Only return new nodes if needed, so if there are no changes, the original object instances are returned.
+    internal abstract class RewritingIRVisitor<TRet, TCtx> : IRNodeVisitor<TRet, TCtx>
+    {
+        protected abstract IntermediateNode Materialize(TRet ret);
+
+        protected abstract TRet Ret(IntermediateNode node);
+
+        public override TRet Visit(TextLiteralNode node, TCtx context)
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/IR/Visitors/RewritingIRVisitor.cs:20]: /// <summary> comments for these methods would be helpful for future implementers.

File: src/libraries/Microsoft.PowerFx.Core/IR/Visitors/RewritingIRVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+                }
+
+                newArgs?.Add(result);
+            }
+
+            return newArgs;
+        }
+
+        // Pass in arg0 is we already called Accept on it.
+        public TRet Visit(CallNode node, TCtx context, TRet arg0)
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/IR/Visitors/RewritingIRVisitor.cs:148]: >public [](http://example.com/codeflow?start=8&length=6)

should this one be private?

Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/IR/Visitors/RewritingIRVisitor.cs:148]: possibly protected? Derived classes need to call it.

File: src/libraries/Microsoft.PowerFx.Core/Public/CheckResult.cs [hunk 3]
```diff
@@ -525,6 +539,19 @@ internal IRResult ApplyIR()
...
+                if (list != null)
+                {
+                    foreach (var transform in list)
+                    {
+                        irnode = transform.Transform(irnode, _errors);
+
+                        // Additional errors from phases. 
+                        // Stop any further processing if we have errors. 
+                        this.ThrowOnErrors();
+                    }
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Public/CheckResult.cs:552]: Wouldn't users generally want all the errors?

File: src/tests/Microsoft.PowerFx.Core.Tests/IRTests/IRPrettyPrinter.cs [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+using Microsoft.PowerFx.Core.IR.Symbols;
+
+namespace Microsoft.PowerFx.Tests
+{
+    // Pretty printer for tests. 
+    // Stable  - tests rely on this, so we need to fix the format 
+    // Compact / No newlines - ideally can fit on a single line in Theorys
+    // Deterministic ordering - when traversing dictionaries
+    // Unambiguous - additional parenthesis as needed to show (a+b)+c vs. a+(b+c). 
+    internal class PrettyPrintIRVisitor : IRNodeVisitor<PrettyPrintIRVisitor.RetVal, PrettyPrintIRVisitor.Context>
```
Comment (lesaltzm) [inline src/tests/Microsoft.PowerFx.Core.Tests/IRTests/IRPrettyPrinter.cs:18]: Would be useful in Core, for CheckResult.PrintIR()

(which, also, should probably not be public on CheckResult??)

**Review by lesaltzm — CHANGES_REQUESTED** [review]
:clock1:

**Review by lesaltzm — APPROVED** [review]
:shipit:

File: src/tests/Microsoft.PowerFx.Core.Tests/IRTests/IRPrettyPrinter.cs (hunk 4 above)
Comment (MikeStall) [reply src/tests/Microsoft.PowerFx.Core.Tests/IRTests/IRPrettyPrinter.cs:18]: starting in Test seemed safe. We could move in future. 
Note that this Compact form is much less invasive than the one in product.   More ideal for test baselines, less ideal for diagnostics.

File: src/libraries/Microsoft.PowerFx.Core/Public/CheckResult.cs (hunk 3 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Public/CheckResult.cs:552]: IR is assuming we have valid input (we don't even try translation if binding fails, see 539). 

So to preserve that invariant, IR is always assuming we're valid, and hence any translation aborts if we get errors.  (warnings are ok). 
It makes it much easier to reason about now - although we could try to expand in the future if we care to get more errors.

**Review by lesaltzm — APPROVED** [review]
🚢

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**